### PR TITLE
Add gemini support

### PR DESCRIPTION
**Pull request by ahmedfahad04**
This pull request introduces support for multiple LLM providers (`openai` and `gemini`) in the `zev` project, along with necessary refactoring and enhancements to handle provider-specific configurations and API interactions. The changes ensure backward compatibility while adding flexibility for users to choose between providers.

### Support for multiple LLM providers:
* Added `DEFAULT_PROVIDER` constant and provider-specific defaults (`DEFAULT_OPENAI_MODEL`, `DEFAULT_GEMINI_MODEL`, etc.) to `src/zev/constants.py`. Backward compatibility is maintained by defaulting to OpenAI settings.
* Introduced `LLM_PROVIDER` as an environment variable and updated `.env` setup prompts to include provider-specific fields for OpenAI and Gemini (e.g., API keys and models) in `src/zev/main.py`.

### Client initialization and API interaction:
* Refactored `get_client` in `src/zev/llm.py` to dynamically return the appropriate client (`openai` or `gemini`) based on the `LLM_PROVIDER` setting. Added `get_openai_client` and `get_gemini_client` helper functions.
* Enhanced `get_options` in `src/zev/llm.py` to handle provider-specific API calls, including JSON parsing and error handling for Gemini responses. Added Gemini-specific prompt formatting and safety settings.

### Codebase refactoring:
* Updated imports and references to constants throughout the codebase to reflect the new provider-specific structure (e.g., `DEFAULT_OPENAI_MODEL` instead of `DEFAULT_MODEL`). [[1]](diffhunk://#diff-4b8303394c15cb875eef5435ab9e035a2f3de192f6a8d419af66713ee5c71ad3L4-L6) [[2]](diffhunk://#diff-695a951e2949ef9673044064a528133628db77bb9ad6d6d2a6afc7d7433a3068L11-R11)

> Would love to know about your feedback!

### CODE REVIEW

File: src/zev/llm.py [hunk 1]
```diff
@@ -49,27 +52,139 @@ class OptionsResponse(BaseModel):
...
+
+Only respond with the JSON format above and nothing else. No markdown formatting, no extra text.
+"""
+            
+            # Configure the model
+            generation_config = {
+                "temperature": 0.1,  # Lower temperature for more deterministic output
+                "top_p": 0.95,
+                "top_k": 40,
+                "response_mime_type": "application/json",  # Hint to return JSON
```
Comment (dtnewman) [inline src/zev/llm.py:131]: I haven't used the Gemini API but with OpenAI it's not just a hint to return json, it is 100% constrained to JSON. Is that possible with Gemini?

File: src/zev/llm.py [hunk 2]
```diff
@@ -49,27 +52,139 @@ class OptionsResponse(BaseModel):
...
+            }
+                
+            safety_settings = [
```
Comment (dtnewman) [inline src/zev/llm.py:134]: What's the thought process on adding these safety settings? I'm not sure that a command line tool needs to worry about this too much and it adds complexity to the code.

Comment (ahmedfahad04) [reply src/zev/llm.py:134]: Yes you are right. I was just trying to follow the best practices, but maybe that type of tight security guardrails should not be imposed here. Thanks for the catch!

File: src/zev/llm.py (hunk 1 above)
Comment (ahmedfahad04) [reply src/zev/llm.py:131]: Initially when I added gemini, it faced difficulties to generated a formated response. But later on, when we tested it seems it works well evenif I don't mention the json output format.

Comment (dtnewman) [reply src/zev/llm.py:131]: Please see https://ai.google.dev/gemini-api/docs/structured-output?lang=python#use-an-enum.

My issue is that with a good prompt, a model will _probably_ return JSON in the format you want. But most of the LLMs now have features that _guarantee_ JSON in a certain structure (or they will return an error) by constraining the logits at each step to the ones that satisfy the constraints.

I think this can be changed to guarantee structured output.

File: src/zev/llm.py (hunk 2 above)
Comment (dtnewman) [reply src/zev/llm.py:134]: It's not that i'm not concerned about safety, i just think:

1) I think it'd be an extreme edge case for users to say "give me a list of commands to find files in the current directory named foo" and it returned a sexually explicit response (even with no filter)

2) There are legitimate uses of sexually explicit responses. For example "give me a command to remove sexually explicit words from this text file and replace with random symbols"

3) As i said earlier, it just adds complexity to the code. Not a lot of complexity in this case, but these things compound over time.

Comment (ahmedfahad04) [reply src/zev/llm.py:134]: I got your points. I have made some adjustments based on your suggestions, will be adding up those changes in the next commit.

File: src/zev/llm.py (hunk 1 above)
Comment (ahmedfahad04) [reply src/zev/llm.py]: I have changed accordingly! Check the latest commits.

File: pyproject.toml [hunk 3]
```diff
@@ -10,7 +10,8 @@ dependencies = [
     "pydantic==2.10.6",
     "python-dotenv==1.0.1",
     "questionary===2.1.0",
-    "rich==13.9.4"
+    "rich==13.9.4",
+    "google-genai==1.11.0"
```
Comment (dtnewman) [inline pyproject.toml:14]: small nitpick, but this should go at the top so that it's in alphabetical order

File: src/zev/llm.py [hunk 4]
```diff
@@ -16,60 +19,87 @@ class OptionsResponse(BaseModel):
...
+                except json.JSONDecodeError as json_err:
+                    print(f"Error: Failed to parse JSON response from Gemini API: {json_err}")
+                    print(f"Raw response: {response_text[:100]}...") 
+                    return None
+
+            except Exception as gemini_err:
+                print(f"Error: Gemini API request failed: {str(gemini_err)}")
+                return None
+                
+    except openai.AuthenticationError:
```
Comment (dtnewman) [inline src/zev/llm.py:99]: i'd break this up into functions with their own try excepts, so you'd have:

```
if provider == "openai":
    return get_open_ai_response()
elif provider == "gemini":
    return get_gemini_response()
else:
    raise Exception(f"Provider {provider} is invalid")

File: src/zev/main.py [hunk 5]
```diff
@@ -22,23 +22,41 @@ class DotEnvField:
 
 
 DOT_ENV_FIELDS = [
+    DotEnvField(
```
Comment (dtnewman) [inline src/zev/main.py:25]: don't like this workflow. If you choose OpenAI as your provider, we should skip asking for the gemini keys and vice versa

File: pyproject.toml (hunk 3 above)
Comment (ahmedfahad04) [reply pyproject.toml:14]: Ops, fixed it.

**Comment by dtnewman** [comment]
Added come more comments. Let me know if you agree and hope you find them helpful :)

File: src/zev/llm.py (hunk 4 above)
Comment (ahmedfahad04) [reply src/zev/llm.py:112]: Done with modularization!

File: src/zev/llm.py [hunk 6]
```diff
@@ -1,75 +1,118 @@
...
+                            
+        try:
+            response_json = json.loads(response.text)
+            return OptionsResponse.parse_obj(response_json)
+        except json.JSONDecodeError as json_err:
+            print(f"Error: Failed to parse JSON response from Gemini API: {json_err}")
+            print(f"Raw response: {response.text[:100]}...") 
+            return None
+
+    except Exception as gemini_err:
```
Comment (dtnewman) [inline src/zev/llm.py:93]: is there a specific Exception type that the Gemini sdk provides for authentication issues?

Comment (ahmedfahad04) [reply src/zev/llm.py:93]: `google_exceptions.InvalidArgument` - this one is an exception primarily raised when we have invalid API key. We have some other exceptions as well like `google_exceptions.Unauthenticated`, `google_exceptions.PermissionDenied` for specific reasons like unauthenticated access or not having enough permission to access specific model.

You can learn more about exception [here](https://googleapis.dev/python/google-api-core/latest/exceptions.html)

**Comment by dtnewman** [comment]
I made some pretty big refactors that will require changes here, but will also make your life a lot easier. See latest `main`.

**Comment by ahmedfahad04** [comment]
I have updated my previous implementation with the necessary refactorings. Thank you for enhancing the module's maintainability through such a flexible code structure, making our lives easier. Let me know if there is anything else I need to refactor.

File: pyproject.toml [hunk 7]
```diff
@@ -4,12 +4,14 @@ version = "0.5.3"
...
-    "questionary>=2.1.0",
-    "rich>=13.9.4"
+    "google-genai==1.11.0",
+    "openai==1.72.0",
+    "platformdirs==4.3.7",
+    "pyperclip==1.9.0",
+    "pydantic==2.10.6",
+    "python-dotenv==1.0.1",
+    "questionary===2.1.0",
+    "rich==13.9.4"
```
Comment (dtnewman) [inline pyproject.toml:14]: let's change back to >=... also we shouldnt need `platformdirs` anymore... i removed that

File: src/zev/main.py (hunk 5 above)
Comment (dtnewman) [reply src/zev/main.py:26]: we can get rid of DOT_ENV_FIELDS and `class DotEnvField:`. I should have taken these out earlier

File: src/zev/llms/gemini/setup.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        name="GEMINI_API_KEY",
+        prompt="Your GEMINI api key:",
+        default="",
+    ),
+    SetupQuestionSelect(
+        name="GEMINI_MODEL",
+        prompt="Choose which model you would like to default to:",
+        options=[
+            SetupQuestionSelectOption(
+            value="gemini-1.5-flash", label="gemini-1.5-flash", description="Low Latency, Good for summarization, Good Performance"
```
Comment (dtnewman) [inline src/zev/llms/gemini/setup.py:14]: nitpick, but i don't like the capitalization. should be 'Low latency, good for summarization, good performance'

File: src/zev/llms/gemini/setup.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+            ),
+            SetupQuestionSelectOption(
+            value="gemini-2.0-flash", label="gemini-2.0-flash", description="Long Context, Good Performance"
```
Comment (dtnewman) [inline src/zev/llms/gemini/setup.py:17]: same here

File: src/zev/constants.py [hunk 10]
```diff
@@ -1,8 +1,21 @@
+class LLMProviders:
+    OPENAI = "openai"
+    OLLAMA = "ollama"
+    GEMINI = "gemini"
+
+
+DEFAULT_PROVIDER = LLMProviders.OPENAI
```
Comment (dtnewman) [inline src/zev/constants.py:7]: we're getting rid of dotenvfields so i dont think this gets used

File: src/zev/constants.py [hunk 11]
```diff
@@ -1,8 +1,21 @@
...
+
+# Default model names for each provider
 OPENAI_DEFAULT_MODEL = "gpt-4o-mini"
+GEMINI_DEFAULT_MODEL = "gemini-2.0-flash"
+
+# Generic default model (fallback)
+DEFAULT_MODEL = OPENAI_DEFAULT_MODEL
```
Comment (dtnewman) [inline src/zev/constants.py:14]: double check, but i think this too

File: src/zev/llms/gemini/provider.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+from zev.config import config
+from google import genai
+from zev.llms.types import OptionsResponse
+from zev.constants import PROMPT
+import json
+
+
+class GeminiProvider(InferenceProvider):
+    def __init__(self):
+        if not config.gemini_model:
```
Comment (dtnewman) [inline src/zev/llms/gemini/provider.py:14]: there should also be a check for api key

File: src/zev/llms/gemini/provider.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+                model=self.model,
+                contents=prompt,
+                config={
+                    'response_mime_type': 'application/json',
+                    'response_schema': OptionsResponse,
+                },
+            )
+            try:
+                response_json = json.loads(response.text)
+                return OptionsResponse.parse_obj(response_json)
```
Comment (dtnewman) [inline src/zev/llms/gemini/provider.py:33]: do we parse the object in openai backend? I think not, and that's probably an oversight on my part

File: src/zev/llms/gemini/provider.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+            except json.JSONDecodeError as json_err:
+                print(f"Error: Failed to parse JSON response from Gemini API: {json_err}")
+                print(f"Raw response: {response.text[:100]}...") 
+                return None
+
+        except Exception as e:
```
Comment (dtnewman) [inline src/zev/llms/gemini/provider.py:40]: i think we spoke about this earlier, but there isn't a specific error for auth failure?

**Comment by dtnewman** [comment]
overall this is trending in the right direction and looks like it'll be mergable soon. I still ahven't actually tested it but i will do so before merging. Good work.

File: src/zev/llms/gemini/provider.py (hunk 14 above)
Comment (ahmedfahad04) [reply src/zev/llms/gemini/provider.py:40]: No, there isn't any specific `AuthFailureException` like that of `OpenAI`.

File: src/zev/config/setup.py [hunk 15]
```diff
@@ -2,6 +2,7 @@
 from pathlib import Path
 import questionary
 from typing import Dict
+from constants import LLMProviders
```
Comment (dtnewman) [inline src/zev/config/setup.py:5]: should be zev.constants

File: src/zev/llms/llm.py [hunk 16]
```diff
@@ -1,13 +1,16 @@
 from zev.config import config
 from zev.llms.openai.provider import OpenAIProvider
 from zev.llms.ollama.provider import OllamaProvider
+from zev.llms.gemini.provider import GeminiProvider
 from zev.llms.inference_provider_base import InferenceProvider
-
+from constants import LLMProviders
```
Comment (dtnewman) [inline src/zev/llms/llm.py:6]: zev.constants

also, nitpick, but can you put it below zev.config import so it's in alphabetical order

File: src/zev/llms/gemini/provider.py (hunk 14 above)
Comment (dtnewman) [reply src/zev/llms/gemini/provider.py:40]: How about this:

```
        except genai.errors.ClientError as e:
            print("Error:", e.details['error']['message'])
            print("Note that to update settings, you can run `zev --setup`.")
            return

```
?

File: src/zev/constants.py (hunk 10 above)
Comment (ahmedfahad04) [reply src/zev/constants.py:7]: I thought using simple strings like "openai" is a bit messy, because we have used that string throughout different scripts. So for some reason, if we need any change, we need to find those values and replace them. Therefore, I tried to replace those strings with a constant type value so that we can make change once and use everywhere.

Let me know your thoughts!

**Comment by dtnewman** [comment]
Yup this is cleaner 😀

File: src/zev/llms/gemini/provider.py (hunk 14 above)
Comment (ahmedfahad04) [reply src/zev/llms/gemini/provider.py:40]: Thanks! That works perfectly

Comment (ahmedfahad04) [reply src/zev/llms/gemini/provider.py]: It's working as expected. Thanks for the carry.

File: src/zev/llms/gemini/provider.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+        self.client = genai.Client(api_key=config.gemini_api_key)
+        self.model = config.gemini_model or GEMINI_DEFAULT_MODEL
+
+    def get_options(self, prompt: str, context: str) -> OptionsResponse | None:
+        try:
+            assembled_prompt = PROMPT.format(prompt=prompt, context=context)
+            response = self.client.models.generate_content(
+                model=self.model,
+                contents=prompt,
```
Comment (dtnewman) [inline src/zev/llms/gemini/provider.py:23]: this needs to be using `assembed_prompt`

**Comment by dtnewman** [comment]
Last thing that needs to happen is the prompt being passed to Gemini is wrong. Needs to be assembled_prompt (this is important!).

File: src/zev/llms/gemini/provider.py (hunk 17 above)
Comment (dtnewman) [reply src/zev/llms/gemini/provider.py:23]: assembled_prompt

File: src/zev/main.py [hunk 18]
```diff
@@ -9,41 +9,11 @@
 import sys
 
 from zev.config.setup import run_setup
-from zev.constants import OPENAI_BASE_URL, OPENAI_DEFAULT_MODEL, CONFIG_FILE_NAME
+from zev.constants import OPENAI_BASE_URL, CONFIG_FILE_NAME, DEFAULT_PROVIDER, GEMINI_DEFAULT_MODEL
```
Comment (dtnewman) [inline src/zev/main.py:12]: some of these are not used... i'll fix after merge

**Review by dtnewman — APPROVED** [review]
Thanks for your hard work :)
